src/main/kotlin/com/devfloor/untitled/comment/application/CommentService.kt

```diff
@@ -1,15 +1,28 @@
 package com.devfloor.untitled.comment.application
 
-import com.devfloor.untitled.article.domain.Article
+import com.devfloor.untitled.article.application.ArticleService
 import com.devfloor.untitled.comment.domain.Comment
 import com.devfloor.untitled.comment.domain.CommentRepository
+import com.devfloor.untitled.common.exception.EntityNotFoundException
+import org.springframework.data.repository.findByIdOrNull
 import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
 
 @Service
 class CommentService(
-    private val commentRepository: CommentRepository,
+    private val repository: CommentRepository,
+    private val articleService: ArticleService,
 ) {
-    fun showAllByArticle(article: Article): List<Comment> {
-        return commentRepository.findAllByArticle(article)
+    private fun showByCommentId(commentId: Long): Comment = repository.findByIdOrNull(commentId)
+        ?: throw EntityNotFoundException("존재하지 않는 댓글입니다.")
+
+    @Transactional
+    fun modifyByArticleId(articleId: Long, commentId: Long, request: CommentModifyRequest) {
+        val article = articleService.showById(articleId)
+        showByCommentId(commentId).apply {
+            modify(
+                Comment(article, this.author, this.anonymous, request.content)
+            )
+        }
```

Comment:
여기 apply를 사용하신 이유가 무엇인가요 !?
이 부분 구현 의도가 `showByCommentId`로부터 comment를 받아, lambda를 실행하기 위한 목적이 맞다면
apply보다는 let이 적절하지 않을까 생각합니다 !! 
https://kotlinlang.org/docs/scope-functions.html#function-selection

++ 추가적으로 스타일적인거라 별로 상관없긴한데, 바꾸시는 김에 라인도 아래와같이 정리하면 깔끔할거같아요!
```suggestion
        showByCommentId(commentId).let {
            modify(Comment(article, this.author, this.anonymous, request.content))
        }
```

Reply:
apply 사용한거 준영이 리뷰 반영한건데 
@lxxjn0  설명 부탁드려요! 저도 궁금해요!

Reply:
음... 제가 생각했을때 (이건 추측입니다.!) 저 메서드 내부에서 `showByCommentId` 를 통해 받은 수신객체인 Comment 객체 자신을 그대로 반환할 것을 의도한거같아요 !
이렇게 하면 저 메서드를 사용한 곳에서는 comment를 그대로 받을 수 있습니다.! 그런데 여기서는 반환이 불필요할거같아요.!

추가적으로 이건 그냥 참고하셔도 좋을거같아서 말씀드립니다 ! apply를 사용하는 경우에는 변수이름이 겹치지 않는 이상 this를 제거해서 `this.author -> author` 요렇게 사용하더라구요 !
그래서 1. 객체를 그대로 반환하고자 할 때, 2. 람다 내부에서 해당 객체 프로퍼티를 초기화할 때 많이 사용하는거같더라구요!?

Reply:
> [Use apply for code blocks that don't return a value and mainly operate on the members of the receiver object.](https://kotlinlang.org/docs/scope-functions.html#apply)

위 링크 참고해보면 좋을 것 같아. 기본적으로 사용하는 용도는 값을 반환하지 않고 설정을 객체 자신에 적용하는 경우에 사용하는 것 같아.

내 생각에 `let`에서는 파라미터로 받는 람다로 T -> R 과 같이 타입을 변경시킬 수 있는 함수를 사용하더라고. scope function에서 R 타입을 Unit으로 지정하면 반환 값이 없어져서 반환값을 없앨 때 사용되기도 하지만 결국 주 목적은 체이닝으로 전달받은 object를 다른 타입을 변경하는게 주 목적인 함수라고 생각해.

타입 변경이 필요없고(전달받은 object를 그대로 사용하고 반환하는 함수) context object의 설정을 변환할 경우 `apply`를 주로 쓰는 것 같아.

내가 생각하기에 지금 상황은 반환값의 유무보다는 기존에 조회한 데이터에 대한 변경이 주요 로직이라고 생각해서 apply를 사용하는게 적절하다고 생각했어. 아 그리고 여기선 this를 쓰면 개인적으로 좀 더 좋을 것 같은게 scope 내부에서 this object의 값만 쓰는게 아니라 외부 변수도 가져와서 쓰고 있어서 this로 구분지어주는게 가독성도 좋고 더 좋을 것 같아.

Reply:
일단 반환값을 고려한게 아니라하니까 그건 둘째치고 너가 준 링크에 나온대로만 얘기해보면
1. `apply`는 주로 수신객체의 멤버들에 대한 작동이고, "블록에서 나오는 다음의 할당을 해당 객체에 적용시켜라"라고 써있어.
2. `let`은 return object를 다른 타입으로 변경하는게 주 목적이 아니라, 1. 체이닝의 결과로 하나 이상의 function을 호출할 때 2. non-null value인 경우에 실행할 때 3. 한정된 스코프 내에서 로컬 변수를 도입할 때 가독성을 위해서. 라고 써있어.

위의 경우에 let의 1번에 해당하는거같고, apply는 수신객체를 초기화할 때 주로 사용한다고 많은 곳에서도 이야기해

Reply:
아 근데 형이 1번에서 말한게 어디 부분이야?
> Use apply for code blocks that don't return a value and mainly operate on the members of the receiver object.

이거 해석하면 '값을 반환하지 않고 수신객체의 맴버에서 작동하는 코드 블록에 apply를 사용한다' 이 말 아닌가?

Reply:
수신객체의 멤버에서 작동하는 코드라는게 그 객체 내부함수나 내부 변수 변경 그런거에 쓰는거 같은데 ?????? 검색 많이 해봣는데 그 객체를 통해 다른 외부 함수로 넘겨서 사용하는 경우 let을 사용하는거같아

Reply:
![image](https://user-images.githubusercontent.com/48052622/119769207-02e8ac80-bef5-11eb-8d8c-dbaa10a9ac3f.png)

2번에서 let의 사용 목적 중 1번이 위 이미지 부분인거 같은데 '체이닝의 결과로 하나 이상의 function을 호출할 때'라면
```kotlin
showByCommentId(commentId).let {
    modify(Comment(article, this.author, this.anonymous, request.content))
    // and more function calls if needed
}
```

이렇게 쓰이는거 말하는 거 아니야? 물론 함수가 하나일 경우도 상관없어서 지금 우리가 사용하는 상황이 여기에 해당된다고 말했다고 이해했어.
그런데 이걸 let을 사용하는 이유로 가정하기에는 사실 scope function이 전부 다 매개변수로 함수를 받는데 이게 let을 사용해야 하는 주된 이유가 되어야 하나 싶긴 해.

개인적으로는 scope function은 람다를 전달받아 실행한다는 것은 전체적으로 거의 비슷하지만, let과 also, run과 apply의 가장 큰 차이는 [문서](https://kotlinlang.org/docs/scope-functions.html#distinctions)에도 이야기하듯 어떻게 수신객체를 참조하는지, 그리고 람다식의 결과를 반환하는 지 컨텍스트 객체를 반환하는 지라고 생각하거든.

내가 scope function들에 대해서 이해한 내용은

- let은 컨텍스트 객체를 활용하여 다양한 처리를 하고 반환타입을 달리 할 수 있다. 이때 컨텍스트 객체는 it으로 호출 가능하다. 그래서 컨텍스트 객체가 가진 메서드가 아니라 다른 함수를 통해 컨텍스트 객체를 처리할 수 있다.
- also는 컨텍스트 객체를 활용하여 다양한 처리를 하고 컨텍스트 객체를 다시 반환할 수 있다. 이때 컨텍스트 객체는 it으로 호출이 가능하다. 컨텍스트 객체가 가진 메서드가 아닌 다른 함수를 사용할 수 있으나, let과는 다르게 처리된 새로운 값은 반환하지 못하고 컨텍스트 객체만 다시 반환이 된다.
- run은 컨텍스트 객체의 맴버(프로퍼티, 메서드)를 활용하여 반환타입을 달리 할 수 있다.  이때 컨텍스트 객체는 this로 호출을 하거나 또는 직접 맴버를 호출할 수 있다. 메서드에서 컨텍스트 객체 자신이 아닌 다른 값을 반환하는 경우 이것으로 반환값을 변경할 수 있다.
- apply는 컨텍스트 객체의 맴버(프로퍼티, 메서드)를 활용하고 컨텍스트 객체를 다시 반환할 수 있다. 이때 컨텍스트 객체는 this로 호출을 하거나 또는 직접 맴버를 호출할 수 있다. 메서드에 반환 타입이 없거나 다르다고 해도 처리 후 컨텍스트 객체를 반환한다.

난 이렇게 전반적으로 이해했거든.

그러면 우리 상황을 생각해보면 위 코드에서의 로직은 'comment 객체를 조회하고 해당 객체를 다른 값으로 수정한다'인데 이때 수정 작업의 경우도 해당 객체(컨텍스트 객체)의 메서드를 사용해서 수행하잖아.
이게 객체의 값을 재설정하는 역할이나 다름없어서 컨텍스트 객체의 메서드를 호출해서 값을 변환하는 지금 상황에서는 apply가 더 적절한 것 같은데 어때?

그리고 위에 제안한 코드가 잘못됐는데, 만약 let으로 고친다면

```kotlin
showByCommentId(commentId).let {
    it.modify(Comment(article, this.author, this.anonymous, request.content))
}
```

이렇게 되는게 맞을 것 같은데 이렇게 쓰는 거보단 apply를 쓰는게 단어 상으로도 코드를 읽을때 '변경을 적용하다'라서 더 적절한 것 같다고 생각해.

Reply:
showByCommentId 를 통해 리턴받은 객체에 대한 내부 modify함수 수행이니까 apply로 써도 될거같은데?
만약 그 객체를 다른 객체?의 함수로 넘겨서 사용할 때 let을 쓰는거 같고..?

Reply:
> 아 근데 형이 1번에서 말한게 어디 부분이야?

Such calls can be read as “ apply the following assignments to the object.” -> "블록에서 나오는 다음의 할당을 해당 객체에 적용시켜라" 여기부분 말하는거야???

저 글 뉘앙스가 수신객체의 멤버들에 대한 일들을 수행할 때나 초기화할때? 사용하는게 적절해보였어. 특히, (T -> R) : T와 같은 상황이 아니라면. (만약 이 상황이면 also, apply 둘중 하나를 사용할텐데 객체 멤버를 이용해 뭔가를 수행해야 하는거면 당연히 also보단 apply가 어울릴테니까)

뭔가 글고 딴데 예시들 봐도 apply를 사용하는건 (위에말한 (T->R):T가 아니면) 주로 객체 상태를 변경하거나 할 때 사용되는거같았고,
> let은 return object를 다른 타입으로 변경하는게 주 목적이 아니라, 1. 체이닝의 결과로 하나 이상의 function을 호출할 때 2. non-null value인 경우에 실행할 때 3. 한정된 스코프 내에서 로컬 변수를 도입할 때 가독성을 위해서. 라고 써있어.

위에 써놓은 이부분에서 일단 위의 상황이 1번에 해당하는건 맞으니까 let이 자연스럽다고 생각했어 !

Reply:
우선은 such로 시작하는 문장에서 내가 앞에서 말해둔 부분이 앞서 나오는데 그 부분을 자르고 뒷 부분만 읽은 것 같아서.

> Use apply for code blocks that don't return a value and mainly operate on the members of the receiver object. The common case for apply is the object configuration. Such calls can be read as “ apply the following assignments to the object.”
>
> **값을 반환하지 않고 주로 수신 객체의 멤버에서 작동하는 코드 블록에서 `apply`를 사용한다.** `apply`의 일반적인 적용 케이스는 객체를 설정하는 부분이다. 이러한 call은 "객체에 다음 할당 적용"으로 읽을 수 있다.

이렇게 해석되는데 우선은 형이 해석한 문장으로는 `블록에서 나오는 다음의 할당을 해당 객체에 적용시켜라` 라고 말했잖아. 그런데 그게 아니고 영어를 읽어보면 `'객체에 매개변수로 전달된 함수를 적용한다'고 읽을 수 있다` 로 해석하는게 맞지 않을까?

그리고 두번째로 (T -> R) : T와 같은 상황에서만 쓰는게 아닌 이유는 첫 문장에서 `값을 반환하지 않고 주로 수신 객체의 멤버에서 작동하는 코드 블록에서 `apply`를 사용한다.`라고 적혀있으니까 설명이 될 것 같아. 우리가 사용하는 부분에서도 Comment 객체의 **메서드(멤버 함수)를 호출**하는 걸 알 수 있고.

형이 말로는 '수신 객체의 맴버(프로퍼티나 메서드)에 대한 일을 수행할 때나 초기화 할 때', 그리고 '주로 객체 상태를 변경할 때' `apply`를 사용하는 것 같았다고 말했는데 지금이 그 상황이지 않아?

지금 comment의 modify, 즉, `comment.modify(Comment(...))` 이걸 쓰고 있는 상황이라서 `apply`가 맞다고 보거든. comment 객체(여기서는 수신 객체)의 멤버에 대한 일(메서드, 프로퍼티의 값을 변경)을 하는 상황이랑 딱 맞아 떨어진다고 생각해서.

근데 혹시 형이 여기서 사용하는 `modify` 함수가 comment의 메서드가 아니고 외부 함수로 본 건 아닐까? 싶은데 만약 그게 아니면 위에 내용들이 내가 생각하는 답변이긴 해. 여기서 사용되는 modify는 comment.kt에 있는 함수잖아.

Reply:
> 1. apply는 주로 수신객체의 멤버들에 대한 작동이고, "블록에서 나오는 다음의 할당을 해당 객체에 적용시켜라"라고 써있어.

아 일단 윗 부분도 읽어서 정리해서 쓴게 이 부분이야 ! "적용시켜라"라고 한건 "apply the following~~~ "이니까 명령문 그대로 쓴거긴해 !

음 일단 내가 이해한거로는 수신객체의 값을 변경하거나 초기화 할때 사용하는게 적절하지 않을까 생각했던것인게, 그걸 이용해서 뭘 해라. 가 아니라 "멤버들에 대한 작동"이라는게 어감이 그렇게 이해됐었어. 실제로 블로그들에서도 많이들 말하는게, 주로 그런 상황에 사용한다고 하기도 하고.

그런데 이 말이 물론 "그 상황이 아니면 사용하지 말아라." 가 아닌건 맞아. 근데, 내가 언급한 말 중 let의 사용용도 1번에 완전히 일치하는 상황이고, 그런 상황이라면 오히려 let이 어웅ㄹ리지 않을까 했던거야.

Reply:
흠.. 어떻게할까욥..